parsely/src/test/java/com/parsely/parselyandroid/ParselyAPIConnectionTest.kt

```diff
@@ -0,0 +1,129 @@
+@LooperMode(LooperMode.Mode.PAUSED)
+class ParselyAPIConnectionTest {
+
+    private lateinit var sut: ParselyAPIConnection
+    private val mockServer = MockWebServer()
+    private val url = mockServer.url("").toString()
+    private val tracker = FakeTracker()
+
+    @Before
+    fun setUp() {
+        sut = ParselyAPIConnection(tracker)
+    }
+
+    @Test
+    fun `given successful response, when making connection without any events, then make GET request`() {
+        // given
+        mockServer.enqueue(MockResponse().setResponseCode(200))
+
+        // when
+        sut.execute(url).get()
+        shadowMainLooper().idle();
+
+        // then
+        val request = mockServer.takeRequest()
+        assertThat(request).satisfies({
+            assertThat(it.method).isEqualTo("GET")
+            assertThat(it.failure).isNull()
+        })
+    }
+
+    @Test
+    fun `given successful response, when making connection with events, then make POST request with JSON Content-Type header`() {
+        // given
+        mockServer.enqueue(MockResponse().setResponseCode(200))
+
+        // when
+        sut.execute(url, pixelPayload).get()
+        shadowMainLooper().idle();
+
+        // then
+        assertThat(mockServer.takeRequest()).satisfies({
+            assertThat(it.method).isEqualTo("POST")
+            assertThat(it.headers["Content-Type"]).isEqualTo("application/json")
+            assertThat(it.body.readUtf8()).isEqualTo(pixelPayload)
+        })
+    }
+
+    @Test
+    fun `given successful response, when request is made, then purge events queue and stop flush timer`() {
+        // given
+        mockServer.enqueue(MockResponse().setResponseCode(200))
+        tracker.events.add(mapOf("idsite" to "example.com"))
+
+        // when
+        sut.execute(url).get()
+        shadowMainLooper().idle();
+
+        // then
+        assertThat(tracker.events).isEmpty()
+        assertThat(tracker.flushTimerStopped).isTrue
+    }
+
+    @Test
+    fun `given unsuccessful response, when request is made, then do not purge events queue and do not stop flush timer`() {
+        // given
+        mockServer.enqueue(MockResponse().setResponseCode(400))
+        val sampleEvents = mapOf("idsite" to "example.com")
+        tracker.events.add(sampleEvents)
+
+        // when
+        sut.execute(url).get()
+        shadowMainLooper().idle();
+
+        // then
+        assertThat(tracker.events).containsExactly(sampleEvents)
+        assertThat(tracker.flushTimerStopped).isFalse
+    }
+
+    @After
+    fun tearDown() {
+        mockServer.shutdown()
+    }
+
+    companion object {
+        val pixelPayload = """
+{
+    "events": [
+        {
+            "idsite": "example.com"
+        },
+        {
+            "idsite": "example2.com"
+        }
+    ]
+}            
+""".trimIndent()
+    }
+
+    private class FakeTracker : ParselyTracker(
+        "siteId", 10, ApplicationProvider.getApplicationContext()
```

Comment:
Initially, I wanted to introduce a package private interface (`QueueManager`?) but due to limitations of Java (interface methods _must_ be `public`) I couldn't do it. Hence, the fake object extending the whole `ParselyTracker`.

Reply:
Praise (❤️): Actually, I like what you did here, so 👍 !